Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import DequeModule
+import NIOConcurrencyHelpers
+
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+/// A protocol for the back-pressure strategy of the ``NIOAsyncSequenceProducer``.
+///
+/// A back-pressure strategy is invoked when new elements are yielded to the sequence or
+/// when a ``NIOAsyncSequenceProducer/AsyncIterator`` requested the next value. The responsibility of the strategy is
+/// to determine whether more elements need to be produced .
+///
+/// If more elements need to be produced, either the ``NIOAsyncSequenceProducerDelegate/produceMore()``
+/// method will be called or a ``NIOAsyncSequenceProducer/Source/YieldResult`` will be returned that indicates
+/// to produce more.
+///
+/// - Important: The methods of this protocol are guaranteed to be called serially. Furthermore, the implementation of these
+/// methods **MUST NOT** do any locking or call out to any other Task/Thread.
+public protocol NIOAsyncSequenceProducerBackPressureStrategy: Sendable {
+    /// This method is called after new elements were yielded by the producer to the source.
+    ///
+    /// - Parameter bufferDepth: The current depth of the internal buffer.
```

Comment:
Can you explain this a bit more? Depth isn't really that self explanatory in this context.

Reply:
Updated the docs to explain this bit more.

> The current depth of the internal buffer of the sequence. The buffer contains all the yielded but not yet consumed elements.